packages/cardano-services/src/Asset/DbSyncAssetProvider.ts

```diff
@@ -0,0 +1,89 @@
+import { Asset, AssetProvider, Cardano, HealthCheckResponse, Provider } from '@cardano-sdk/core';
+import {
+  CardanoTokenRegistry,
+  CardanoTokenRegistryConfiguration,
+  CardanoTokenRegistryDependencies
+} from './CardanoTokenRegistry';
+import { DbSyncNftMetadataService, DbSyncNftMetadataServiceDependencies } from './DbSyncNftMetadataService';
+import { NftMetadataService, TokenMetadataService } from './types';
+import { createDbSyncMetadataService } from '../Metadata';
+
+/**
+ * Configuration options for DbSyncAssetProvider
+ */
+export type DbSyncAssetProviderConfiguration = CardanoTokenRegistryConfiguration;
+
+/**
+ * Dependencies that are need to create DbSyncAssetProvider
+ */
+export type DbSyncAssetProviderDependencies = {
+  /**
+   * The NftMetadataService to retrieve Asset.NftMetadata .
+   * Default: new DbSyncNftMetadataService with dependencies taken from DbSyncAssetProviderDependencies
+   */
+  ntfMetadataService?: NftMetadataService;
+
+  /**
+   * The TokenMetadataService toretrieve Asset.TokenMetadata .
+   * Default: new CardanoTokenRegistry with dependencies taken from DbSyncAssetProviderDependencies
+   * and configuration from DbSyncAssetProviderConfiguration
+   */
+  tokenMetadataService?: TokenMetadataService;
+} & CardanoTokenRegistryDependencies &
+  Pick<DbSyncNftMetadataServiceDependencies, 'logger' | 'db'> &
+  Partial<Pick<DbSyncNftMetadataServiceDependencies, 'metadataService'>>;
+
+interface DbSyncAssetProviderDependenciesWithRequired extends DbSyncAssetProviderDependencies {
+  ntfMetadataService: NftMetadataService;
+  tokenMetadataService: TokenMetadataService;
+}
+
+type AssetExtraData = Parameters<AssetProvider['getAsset']>[1];
+
+/**
+ * AssetProvider implementation using  CardanoTokenRegistry, DbSyncNftMetadataService
+ * and cardano-db-sync database as sources
+ */
+export class DbSyncAssetProvider implements AssetProvider, Provider {
+  #deps: DbSyncAssetProviderDependenciesWithRequired;
+
+  constructor(dependencies: DbSyncAssetProviderDependencies, config: DbSyncAssetProviderConfiguration = {}) {
+    const { db, logger } = dependencies;
+
+    this.#deps = {
+      ntfMetadataService: new DbSyncNftMetadataService({
+        metadataService: createDbSyncMetadataService(db, logger),
+        ...dependencies
+      }),
+      tokenMetadataService: new CardanoTokenRegistry(dependencies, config),
+      ...dependencies
+    };
+  }
+
+  async close() {
+    this.#deps.tokenMetadataService.shutdown();
+  }
+
+  async getAsset(id: Cardano.AssetId, extraData?: AssetExtraData) {
+    // TODO SELECT FROM multi_asset
```

Comment:
Hi @rhyslbw and @mkazlauskas , I open this PR prematurely to ask for your support.

Looking at [`DbSyncNftMetadataService`](https://github.com/input-output-hk/cardano-js-sdk/blob/feat/ADP-2029-asset-provider/packages/cardano-services/src/Asset/DbSyncNftMetadataService.ts) I think I can get an idea on how to perform queries on db-sync.

My problem here is not how to perform a query, is what query I have to do. From [this comment](https://github.com/input-output-hk/cardano-js-sdk/pull/321#issuecomment-1180300713) I understood I have to perform a `SELECT FROM  multi_asset`, but I have in input a [`Cardano.AssetId`](https://github.com/input-output-hk/cardano-js-sdk/blob/feat/ADP-2029-asset-provider/packages/core/src/Provider/AssetProvider/types.ts#L14) and I don't know how to convert it in a `id` for the `multi_asset` table or in an [`AssetPolicyIdAndName`](https://github.com/input-output-hk/cardano-js-sdk/blob/feat/ADP-2029-asset-provider/packages/cardano-services/src/Asset/types.ts#L7) in order to perform the query.

I ask for your support as ATM this is a blocking point for me.

Thank you in advance

Reply:
The `id` from the `multi_asset` table is auto generated by postgres so we can't use that with the `AssetId`.

The `AssetId` is a concatenation of the asset policy and name, which you can get by using these functions here https://github.com/input-output-hk/cardano-js-sdk/blob/master/packages/core/src/Asset/util/assetId.ts#L4

Once you have that you can use them in the SQL queries. You might need to convert them to a `Buffer` before with https://github.com/input-output-hk/cardano-js-sdk/blob/master/packages/cardano-services/src/util/hexString.ts#L2 but I'm not totally sure